File: fsharp/hangman/Hangman.fs
module Hangman
open System

type Progress = 
    | Busy of int
    | Win
    | Lose

type State = { progress: Progress; maskedWord: string }

type HangmanGame (word:string) =

    let word = word |> Seq.toList
    let guess = new Event<char>()
    let stateChange = new Event<State>()

    let initialState = { 
        maskedWord = word |> Seq.map (fun _ -> '_') |> String.Concat; 
        progress = Busy 9 
    }

    let advanceGame (state:State) (c:char) =
        match state.progress with
        | Win | Lose -> state
        | Busy n ->
            if state.maskedWord.Contains (string c) || Seq.contains c word |> not then
                let newProgress = n - 1
                { state with progress = if newProgress = 0 then Lose else Busy newProgress }
            else
                let indexes = [0..word.Length - 1] |> List.filter (fun i -> word.[i] = c)
                let newWord = 
                    state.maskedWord 
                    |> Seq.mapi (fun i cs -> if List.contains i indexes then c else cs) 
                    |> String.Concat
                if newWord.Contains "_" then
                    { state with maskedWord = newWord }
                else
                    { state with maskedWord = newWord; progress = Win }

    member _this.Guess c = guess.Trigger c
    member _this.StateChange = stateChange.Publish

    member _this.Start = 
        stateChange.Trigger initialState
        guess.Publish
        |> Event.scan advanceGame initialState
        |> Observable.add (stateChange.Trigger)

let createGame word = new HangmanGame (word)

let statesObservable (game: HangmanGame) = game.StateChange

let startGame (game: HangmanGame) = game.Start

let makeGuess c (game: HangmanGame) = game.Guess c

File: csharp/hangman/Hangman.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class HangmanState
{
    public HangmanStatus Status { get; set; }
    public int RemainingGuesses { get; set; }
    public string MaskedWord { get; set; }
    public HashSet<char> Guesses { get; set; }
}

public enum HangmanStatus
{
    Busy, Win, Lose
}

public class HangmanGame
{
    public event EventHandler<HangmanState> StateChanged;

    private readonly string unmaskedWord;
    private HangmanState state;

    public HangmanGame(string word) => unmaskedWord = word;

    public void Start()
    {
        state = new HangmanState
        {
            Status = HangmanStatus.Busy,
            RemainingGuesses = 9,
            MaskedWord = new string('_', unmaskedWord.Length),
            Guesses = new HashSet<char>()
        };
        StateChanged?.Invoke(this, state);
    }

    public void Guess(char c)
    {
        if(state.Status != HangmanStatus.Busy)
            return;

        state.Guesses.Add(c);
        if(state.MaskedWord.Contains(c) || !unmaskedWord.Contains(c))
            if(state.RemainingGuesses == 0)
                state.Status = HangmanStatus.Lose;
            else
                state.RemainingGuesses --;
        else
            state.MaskedWord = new string(unmaskedWord.Select(oc => 
                state.Guesses.Contains(oc) ? oc : '_').ToArray());

        if(!state.MaskedWord.Contains('_'))
            state.Status = HangmanStatus.Win;

        StateChanged?.Invoke(this, state);
    }
}